docs/releasing.md

```diff
@@ -0,0 +1,73 @@
+# Releasing
+
+## Prerequisites
+
+### `docker`
+
+You must have docker installed.
+
+## Output
+
+### Expected artifacts
+
+1. A container image of the shared baremetal-operator
+1. A git tag
+1. A release on Github containing:
+    - A manifest file - `infrastructure-components.yaml`
+    - A metadata file - `metadata.yaml`
+    - A cluster template - `cluster-template.yaml`
+
+### Artifact locations
+
+1. The container image is found in the registry `quay.io/metal3-io` with an image
+   name of `baremetal-operator` and a tag that matches the release
+   version. The image is automatically built once the release has been created.
+
+## Creating a release for BMO
+
+### Process
+
+For version v0.x.y:
+
+1. Create the release notes `make release-notes`. Copy the output and sort
+   manually the items that need to be sorted.
+1. Create an annotated tag `git tag -a v0.x.y -m v0.x.y`. To use your GPG
+   signature when pushing the tag, use `git tag -s [...]` instead
+1. Push the tag to the GitHub repository `git push origin v0.x.y`
+   NB: `origin` should be the name of the remote pointing to
+   `github.com/metal3-io/baremetal-operator`
+1. [Create a release in GitHub](https://help.github.com/en/github/administering-a-repository/creating-releases)
+   that contains the elements listed above that have been created in the `out`
+   folder
+1. Create a branch `release-0.x` for a minor release for backports and bug fixes.
+
+### Permissions
+
+Releasing requires a particular set of permissions.
+
+- Tag push access to the GitHub repository
+- GitHub Release creation access
+
+## Impact on Metal3
+
+Multiple additional actions are required in the Metal3 project
+
+### Update the Jenkins jobs
+
+For each minor or major release, two jobs need to be created :
+
+- a master job that runs on a regular basis
+- a PR verification job that is triggered by a keyword on a PR targeted for that
+  release branch.
```

Comment:
we will need to discuss how we deal with the BMO versions in the tests, if we will need multiple jobs etc. and the compatibility matrix with CAPM3 to cover it with the tests.

Reply:
It seems like CAPM3 can control that. We would want test jobs for each CAPM3 branch against the head of the branch for the series from BMO that is supported. We really only need to support one version of BMO per version of CAPM3, right? So I don't think that we would end up with a lot more jobs than we have now. The BMO is at the bottom of the dependency stack, so if we cross-test, so that changes to BMO can't break CAPM3 jobs, then we shouldn't need CAPM3 jobs using pinned versions of BMO.

Reply:
I guess that would depend on how we consider BMOfrom CAPM3 point of view, if it is just a dependency updated from time to time, allowing complete split of development cycles, then we'd need separate testing for BMO, considering that CAPM3 might not support the new version / feature yet. But if we go for synced development cycles, then a version or feature change in BMO would mean a sync change in CAPM3. That's how we have been doing it until now and probably the easiest with regards to the testing, allowing for full integration tests at any point in BMO development cycles. However, going this way probably means that we need to release BMO for every CAPM3 release.

Reply:
I think we're saying similar things.

I envision a branch in BMO for each CAPM3 release. That BMO branch would be tested with the corresponding CAPM3 branch, so that the "series" of releases from the branches stay compatible.

We would do the same for the master branch of each component, to ensure that upcoming releases are compatible. 

Because the BMO is at the bottom of the dependency stack, there may be releases of BMO on a branch that do not correspond to a new CAPM3 release. CAPM3 could choose to consume those explicitly, or not, as long as future changes on each BMO branch continue to be compatible with the same branch of CAPM3. That doesn't really seem very different from what we're doing today, where CAPM3 always consumes the image created from the tip of the branch implicitly.

That leaves the issue of upgrade testing between branches. I think we want to reach a point of stability where we ensure compatibility between series for upgrades, and that would require another test job to install a cluster with one version, update the components, and then ensure that the upgraded cluster stabilizes and there are no errors or data loss. The design and implementation of that test job will be influenced by the release plan, but I think we should come up with the plan for that separately over in metal3-docs.

Reply:
I agree. I am thinking about what would happen when there would need to be a new minor or major release of BMO with breaking change. I guess that would force us to do a similar minor/major release on CAPM3. It might be what we want , because such an API change would need to be visible for the consumer of CAPM3 that still create the BMHs. So this synchronization approach could work as long as we ensure that major changes in BMO are reflected in CAPM3, and we don't hide a breaking change of BMO in CAPM3.

For the upgrade tests, I guess it would be done anyways as we are doing the CAPM3 upgrade tests, since an upgrade of BMO would mean an upgrade of CAPM3.

So it looks like the best option for us is to keep the two release cycles somewhat in sync.

Reply:
Yes, I think if we're going to have a breaking API change in BMO we would want to have a release of both components before that happens. Ideally we will be able to make API changes compatible by using web hooks, which would let us make the change in BMO first and then update CAPM3 to require the newer API later.